### PR TITLE
Allow forcing HTTP Connect proxy passthrough for Konnectivity support

### PR DESCRIPTION
**Pull request by steeling**
Konnectivity server only accepts HTTP CONNECT passthrough proxying. The standard lib proxies via a GET request for destinations that are non-TLS however. This means that when scraping a non-secured HTTP endpoint through a proxy, the request is always sent as a GET. Only HTTPS destinations are sent as a CONNECT Passthrough. There is nothing in any spec that requires this however. 

This PR adds a flag to scrape configs, to force HTTP destinations to use CONNECT passthrough.

See #5109

### CODE REVIEW

File: lib/promscrape/client.go
Comment (zekker6) [inline lib/promscrape/client.go:99]: Could you please place a full link to an issue here?
It is usually easier to follow the link later on.

File: lib/promscrape/scrapework.go
Comment (zekker6) [inline lib/promscrape/scrapework.go:110]: Is this variable needed here or only one defined in `ProxyClientConfig` is needed?

**Review by zekker6 — COMMENTED** [review]
Hello @steeling, thank you for the PR!

Please, see a few comments.

Also, to get this merged it is required to update the SD section of documentation here - https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/docs/sd_configs.md
Could you please add description of new parameter there?

File: lib/promscrape/scrapework.go
Comment (steeling) [reply lib/promscrape/scrapework.go:110]: So `ScrapeConfig` has a `ProxyClientConfig` which is a natural place to put this. However that config gets translated to a generic struct `promauth.Config` on both `ScrapeWorkConfig` and `ScrapeWork`.

The `promauth.Config` is used for both destination endpoints and proxying, and is just TLS info. Adding it there would leave a dangling appendage on the regular `AuthConfig` field.

That said, I'm open to doing whatever the VM team recommends :)

File: lib/promscrape/client.go
Comment (steeling) [reply lib/promscrape/client.go]: done!

**Comment by steeling** [comment]
> Hello @steeling, thank you for the PR!
> 
> Please, see a few comments.
> 
> Also, to get this merged it is required to update the SD section of documentation here - https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/docs/sd_configs.md Could you please add description of new parameter there?

done, thanks for the feedback!

File: lib/promscrape/scrapework.go
Comment (zekker6) [reply lib/promscrape/scrapework.go:110]: Thank you for the details, that makes sense to me.

Could you please add a comment for this parameter to match the code style?

Comment (steeling) [reply lib/promscrape/scrapework.go:111]: done!

File: docs/sd_configs.md
Comment (zekker6) [inline docs/sd_configs.md:1614]:
```suggestion
    # proxy_force_http_connect: <boolean>
```

File: lib/promscrape/client_test.go
Comment (zekker6) [inline lib/promscrape/client_test.go:18]: Could you please remove `testify` dependency?
We're usually writing tests with standard Go functions and just making comparisons manually, you can refer to these tests for example: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/promscrape/config_test.go

**Review by zekker6 — COMMENTED** [review]
See a few comments, LGTM other than that.

File: lib/promscrape/client_test.go
Comment (steeling) [reply lib/promscrape/client_test.go]: done!

File: docs/sd_configs.md
Comment (steeling) [reply docs/sd_configs.md]: done!

**Review by zekker6 — APPROVED** [review]

**Comment by steeling** [comment]
@zekker6 PTAL again, in testing this I noticed I hadn't implemented the fix for configs that set `stream_parse` to true. I've implemented this now, and added some more robust tests.

**Comment by steeling** [comment]
Hi all, friendly ping on this.

cc: @zekker6, @hagen1778, @f41gh7

**Comment by hagen1778** [comment]
@zekker6 I have a little knowledge of this. @f41gh7 could you please take a look? I remember you did something similar already.

**Comment by zekker6** [comment]
@steeling Sorry for the delay, I've completely missed this. Will take a look either later today or on Monday.

**Review by zekker6 — APPROVED** [review]
LGTM

**Comment by steeling** [comment]
thanks @zekker6

Pushed fixes for lint, and signed all commits

**Comment by steeling** [comment]
@zekker6, can you help get reviews from remaining approvers?

cc: @hagen1778 , @f41gh7

**Comment by steeling** [comment]
Hey folks, if I continued on this PR could we get this in?

**Comment by zekker6** [comment]
Hey @steeling, let me ping other maintainers internally to get the approvals.

> Hey folks, if I continued on this PR could we get this in?

I don't see why not, so it would be great if you could update the PR to get this merged.

**Comment by hagen1778** [comment]
The feature looks good to me. But I need technical approvement from @f41gh7 for implementation.

File: lib/promscrape/client.go
Comment (f41gh7) [reply lib/promscrape/client.go:129]: for instance, looks like proxyURL = "socks5://some-proxy" could fail. It'd be great to have test for it. But it maybe too complicated.

File: lib/promscrape/client.go
Comment (f41gh7) [inline lib/promscrape/client.go:157]: Looks like it's possible to use  single `dialFunc` instead of `streamDialFunc`.

**Review by f41gh7 — COMMENTED** [review]
Thanks overall looks good. Have a few comments.

Also, it'd be great to have a changelog entry with proposed changes.

**Comment by f41gh7** [comment]
Sorry, I just realized that given PR is outdated. 

There was a major change into master branch, that remove fasthttp client and it makes impossible to merge current PR with it.  And it will requires a significant amount of work and changes to add "CONNECT method" to current branch.

 I see the following options:
1) address issue to the https://github.com/kubernetes-sigs/apiserver-network-proxy . Their proxy server is incompatible with go standard library http/transport. 
2) try to rewrite patch to the vmagent. It'll require changes to the dial func https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/netutil/statdial.go#L15 and https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/promscrape/client.go#L64


 Lets continue communication at linked issue.

**Pull request closed without merging**
